magma/lib/magma/query/question.rb

```diff
@@ -168,72 +189,115 @@ def count_query
       end
 
       query.select(
-        *@start_predicate.select
+          *order_by_column_names.zip(order_by_aliases).map { |c, a| c.as(a) }
       )
     end
 
     # get page bounds for this question using @options[:page] and @options[:page_size]
     def bounds_query
       raise QuestionError, 'Page size must be greater than 1' unless @options[:page_size] > 1
+      bounds_select_parts = order_by_aliases.dup
+      bounds_select_parts << Sequel.function(:row_number)
+                             .over(order: order_by_aliases)
+                             .as(:row)
+
       count_query.from_self.select(
-        # add row_numbers to the count query
-        @start_predicate.identity,
-        Sequel.function(:row_number)
-          .over(order: @start_predicate.identity)
-          .as(:row)
+          *bounds_select_parts
       ).from_self(alias: :main_query).select(
-        # only the first row from each page
-        @start_predicate.identity).where(
+          # only the first row from each page
+          *order_by_aliases).where(
           Sequel.lit(
-            '? % ? = 1',
-            Sequel[:main_query][:row],
-            @options[:page_size]
+              '? % ? = 1',
+              Sequel[:main_query][:row],
+              @options[:page_size]
           )
-        )
+      )
     end
 
     def page_bounds_query
-      bounds_query.limit(2).offset(@options[:page]-1)
+      bounds_query.limit(2).offset(@options[:page] - 1)
     end
 
     # create an upper and lower limit for each page bound
     def all_bounds
       bounds = to_table(bounds_query)
       bounds.map.with_index do |row, index|
-        bound = {:lower =>  row[@start_predicate.identity], :upper => nil}
+        bound = {:lower => order_by_aliases.map { |c| row[c] }, :upper => nil}
 
         if bounds[index + 1]
-          bound[:upper] = bounds[index + 1][@start_predicate.identity]
+          bound[:upper] = bounds[index + 1].map { |c| row[c] }
         end
 
         bound
       end
     end
 
+    def apply_multi_stage_ordering_bounds(query, upper: nil, lower: nil)
```

Comment:
Now that bounds is a tuple of potential multiple values (like an ordering on updated_at, id), we need a more correct way of selecting values between multi-column intervals.  This is not as simple as selecting where(a >= 1 & b >= 2), since there is no guarantee the total ordering of a and b coincide.  Instead, we need to select where for each element of the tuple, either being  greater or equal to  the lower bound && either each additional tuple element being greater or equal, or the previous tuple elements being strictly greater.  

The logic is kind of noted in the comments, it's a tricky looking algorithm but works.